src/HttpClientPoolItem.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Http\Client\Common;
+
+use Psr\Http\Message\RequestInterface;
+use Http\Client\Exception;
+
+/**
+ * A HttpClientPoolItem represent a HttpClient inside a Pool.
+ *
+ * It is disabled when a request failed and can be reenable after a certain number of seconds
+ * It also keep tracks of the current number of request the client is currently sending (only usable for async method)
+ *
+ * @author Joel Wurtz <joel.wurtz@gmail.com>
+ */
+class HttpClientPoolItem extends FlexibleHttpClient
+{
+    /** @var int Number of request this client is actually sending */
+    private $sendingRequestCount = 0;
+
+    /** @var bool Status of the http client */
+    private $disabled = false;
+
+    /** @var \DateTime Time when this client has been disabled */
+    private $disabledAt;
+
+    /** @var int|null Number of seconds after this client is reenable, by default null: never reenable this client */
+    private $reenableAfter;
+
+    /**
+     * {@inheritdoc}
+     *
+     * @param null|int $reenableAfter Number of seconds after this client is reenable
+     */
+    public function __construct($client, $reenableAfter = null)
+    {
+        parent::__construct($client);
+
+        $this->reenableAfter = $reenableAfter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function sendRequest(RequestInterface $request)
+    {
+        if ($this->isDisabled()) {
+            throw new Exception\RequestException('Cannot send request as this client as been disabled', $request);
+        }
+
+        try {
+            ++$this->sendingRequestCount;
+            $response = parent::sendRequest($request);
+            --$this->sendingRequestCount;
+        } catch (Exception $e) {
+            $this->disable();
+            --$this->sendingRequestCount;
+
+            throw $e;
+        }
+
+        return $response;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function sendAsyncRequest(RequestInterface $request)
+    {
+        if ($this->isDisabled()) {
+            throw new Exception\RequestException('Cannot send request as this client as been disabled', $request);
+        }
+
+        ++$this->sendingRequestCount;
+
+        return parent::sendAsyncRequest($request)->then(function ($response) {
+            --$this->sendingRequestCount;
+
+            return $response;
+        }, function ($exception) {
+            $this->disable();
```

Comment:
ah, found the reenable. but still, is your assumption that one client stops working while others continue to work? as this is not multithread or multiprocess code, i don't expect this to be a relevant use case.

or is the idea to use a bunch of clients to do round robin on the client to different instances of the same server application rather than on a load balancer in front of the servers?